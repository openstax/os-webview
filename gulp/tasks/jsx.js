const path = require('path');
const gulp = require('gulp');
const config = require('../config');
const pi = require('gulp-load-plugins')({
    pattern: ['gulp-*', 'gulp.*', 'del']
});

function jsx() {
    const configSrcApp = `${config.src}/app`;
    const configDestApp = `${config.dest}/app`;

    return gulp.src(`${configSrcApp}/**/*.jsx`, {
        since: gulp.lastRun(jsx)
    })
    .pipe(pi.sourcemaps.init({loadMaps: true}))
    .pipe(pi.rename((uri) => {
        uri.extname = '.jsx.js';
    }))
    .pipe(pi.babel({
        compact: false,
        presets: ['@babel/preset-react']
    }))
    // prefix the sourcemaps with '../src/' so webpack can find them
    .pipe(pi.sourcemaps.mapSources(function(sourcePath, file) {
      const sourcePathHtml = sourcePath.replace('.jsx.js', '.jsx');
      const rel = path.relative(path.dirname(`${configDestApp}/${sourcePath}`), `${configSrcApp}/${sourcePathHtml}`)
      return rel;
    }))
    .pipe(pi.sourcemaps.write('.'))
    .pipe(gulp.dest(`${configDestApp}`));
}


function watchJsx() {
    gulp.watch([`${config.src}/**/*.jsx`], config.watchOpts, gulp.series(
        jsx
    ));
}

exports.jsx = jsx;
exports.jsx.watch = watchJsx;
