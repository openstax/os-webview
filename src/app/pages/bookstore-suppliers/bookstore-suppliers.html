<template args="model">
    <div if="model.headline" class="hero">
        <div class="content">
            <div class="text">
                <h1>{model.headline}</h1>
                <div data-html="subhead" skip="true"></div>
                <div class="larger-screen" data-html="subhead2" skip="true"></div>
            </div>
        </div>
        <div class="images">
            <img class="desktop-bg" src="/images/bookstore-suppliers/hero-bookstore-print-desktop.png">
        </div>
    </div>
    <div if="model.suppliers" class="main-content">
        <div class="text-section">
            <h2>Providers</h2>
            <div class="intro-blurb">{model.featuredSuppliersBlurb}</div>
        </div>
        <div each="supplier in model.featuredSupplier" class="card featured">
            <div class="logo-dot">
                <img src="{supplier.logoUrl}" alt="company logo">
            </div>
            <h2>{supplier.name}</h2>
            <div class="blurb" data-html="featuredSupplier[{$item}].description" skip="true"></div>
            <a class="btn primary" href="{supplier.buttonUrl}">{supplier.buttonText}</a>
        </div>
        <div class="intro-blurb">{model.suppliersBlurb}</div>
        <div class="cards">
            <div each="supplier in model.suppliers" class="card">
                <div if="supplier.isCanadian" class="flag-card">
                    <div class="canada-flag">
                        <div class="white-field">
                            <i class="fab fa-canadian-maple-leaf"></i>
                        </div>
                    </div>
                </div>
                <div class="logo-dot">
                    <img src="{supplier.logoUrl}" alt="company logo">
                </div>
                <h2>{supplier.name}</h2>
                <div class="blurb" data-html="suppliers[{$item}].description" skip="true"></div>
                <a class="btn primary" href="{supplier.buttonUrl}">{supplier.buttonText}</a>
            </div>
        </div>
        <h2>Price lists</h2>
        <div class="split-card">
            <div>
                {model.usButtonText}
                <a href="{model.usButtonUrl}" class="btn primary">
                    View <i class="fas fa-external-link-alt"></i>
                </a>
            </div>
            <div>
                {model.caButtonText}
                <a href="{model.caButtonUrl}" class="btn primary">
                    View <i class="fas fa-external-link-alt"></i>
                </a>
            </div>
        </div>
    </div>
</template>
