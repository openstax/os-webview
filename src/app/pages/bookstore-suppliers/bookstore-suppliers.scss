@import 'pattern-library/core/pattern-library/headers';
@import 'mixins/placeholder-selectors';

.bookstore-suppliers {

    @include width-up-to($phone-max) {
        .larger-screen {
            display: none;
        }
    }

    @include wider-than($phone-max) {
        .small-screen {
            display: none;
        }
    }

    .hero {
        @extend %light-on-dark;

        background-image: linear-gradient(to top right, os-color(light-blue), os-color(blue));
        color: text-color(white);
        overflow: visible;
        padding-left: 0;
        padding-right: 0;
        z-index: 2;

        .content {
            @extend %content;
        }

        .text {
            display: grid;
            grid-gap: $normal-margin;
            position: relative;
            text-align: left;
            z-index: 1;

            @include wider-than($phone-max) {
                @include set-font(body-large);

                max-width: 55rem;
                width: 50%;
            }

            h1 {
                @include set-font(h1);
            }

            a {
                color: inherit;
            }
        }

        .images {
            overflow: visible;
            position: relative;
            width: 100%;

            .desktop-bg {
                display: none;

                @include wider-than($phone-max) {
                    // Bottom and right have to overcome the padding of hero
                    bottom: -6rem;
                    display: block;
                    max-height: 30rem;
                    position: absolute;
                    right: -1.5rem;
                }
            }
        }
    }

    .main-content {
        @extend %content;

        display: grid;
        grid-gap: 6rem;
        justify-content: center;
        z-index: 2;

        @include width-up-to($phone-max) {
            background-color: ui-color(white);
            grid-gap: 4rem;
            padding-bottom: $normal-margin;
            padding-top: 3rem;
            z-index: 1;

            [data-html="subhead2"] {
                margin-bottom: -3rem;
            }
        }

        @include width-between($phone-max, $tablet-max) {
            padding: 4rem $normal-margin;
        }

        @include wider-than($tablet-max) {
            padding-bottom: 8rem;
            padding-top: 6rem;
        }

        > * {
            margin: 0;
        }

        .text-section {
            @include set-font(body-large);

            display: grid;
            grid-gap: 3rem;

            > * {
                margin: 0;
            }
        }

        button {
            @extend %button;
            max-width: calc(100vw - 3rem);

            &.primary {
                @extend %primary;
            }
        }
    }

    .button-row {
        display: grid;
        grid-auto-flow: column;
        grid-column-gap: 3rem;
        justify-content: center;
    }

    .btn {
        @extend %button;
        @extend %secondary;

        justify-self: center;

        &.primary {
            @extend %primary;
        }
    }

    .intro-blurb {
        @include set-font(body-large);
    }

    .swipable-row {
        @include width-up-to($phone-max) {
            width: 100%;
            overflow-x: auto;
        }
    }

    .cards {
        background-color: ui-color(default);
        display: grid;
        grid-gap: 3rem;
        margin-left: -$normal-margin;
        margin-right: -$normal-margin;
        overflow-x: auto;
        padding: 0 $normal-margin;

        $narrow-tablet: 78em * $scale-factor;

        @include width-up-to($phone-max) {
            background-color: ui-color(white);
            grid-auto-flow: column;

            .card {
                min-width: 28rem;
            }
        }

        @include width-between($phone-max, $narrow-tablet) {
            grid-template-columns: repeat(2, 1fr);
        }

        @include width-between($narrow-tablet, $media-content-max) {
            grid-template-columns: repeat(3, 1fr);
        }

        @include wider-than($media-content-max) {
            grid-template-columns: repeat(4, 1fr);
        }
    }

    .card {
        @extend %card;

        background-color: ui-color(white);
        display: grid;
        grid-gap: 2rem;
        grid-template-rows: auto auto 1fr auto;
        justify-items: center;
        margin-bottom: 0.3rem;
        margin-top: 0.3rem;
        overflow: hidden;
        padding: 3rem;
        position: relative;
        text-align: center;

        $canada-red: #f00;
        .flag-card {
            @extend %card;

            border-radius: 0.2rem;
            padding: 0.9rem 1rem;
            position: absolute;
            right: 3rem;
            top: 0;

            .canada-flag {
                background-color: $canada-red;
                display: grid;
                justify-content: center;
                width: 4.2rem;

                .white-field {
                    @include set-font(body-large);

                    align-items: center;
                    background-color: ui-color(white);
                    color: $canada-red;
                    display: flex;
                    height: 2.2rem;
                    justify-content: center;
                    width: 2.2rem;
                }
            }
        }

        @include wider-than($phone-max) {
            &.featured {
                grid-column-gap: 3rem;
                grid-template: 'icon headline'
                'icon blurb'
                'icon button';
                justify-items: left;
                text-align: left;

                > .logo-dot {
                    grid-area: icon;
                    height: 30rem;
                    width: 30rem;
                }

                > h2 {
                    grid-area: headline;
                }

                > .blurb {
                    @include body-font(2.4rem);

                    grid-area: blurb;
                }

                > .btn {
                    grid-area: button;
                    justify-self: left;
                }
            }
        }

        &:not(.featured) h2 {
            @include title-font(2.4rem);

            line-height: normal;
            margin: 0;
        }

        .logo-dot {
            align-items: center;
            background-color: ui-color(default);
            border-radius: 50%;
            display: flex;
            height: 21rem;
            justify-content: center;
            margin: 1rem 0;
            width: 21rem;

            @include width-between($phone-max, $tablet-max) {
                height: 17rem;
                width: 17rem;
            }

            > img {
                max-height: 80%;
                max-width: 80%;
            }
        }

        .btn {
            height: 5rem;
            max-width: 100%;
            overflow: hidden;

            @include width-between($phone-max, $tablet-max) {
                padding-left: 0;
                padding-right: 0;
                width: 100%;
            }
        }

        ~ h2 {
            @include wider-than($phone-max) {
                margin-bottom: -3rem;
                margin-top: 3rem;
            }
        }
    }

    .split-card {
        @extend %card;

        background-color: ui-color(white);
        display: grid;
        max-width: 57rem;
        padding: 0 3rem;

        > div {
            align-items: center;
            display: grid;
            grid-column-gap: 3rem;
            grid-template-columns: 1fr auto;
            padding: 3rem 0;
            width: 100%;

            &:not(:first-child) {
                border-top: thin solid ui-color(form-border);
            }

            .btn {
                height: 3rem;
                padding-left: $normal-margin;
                padding-right: $normal-margin;

                .fa-external-link-alt {
                    margin-left: 0.7rem;
                }
            }
        }
    }
}
