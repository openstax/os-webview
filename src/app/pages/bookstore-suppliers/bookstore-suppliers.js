import {pageWrapper} from '~/controllers/jsx-wrapper';
import React, {useState, useEffect} from 'react';
import {usePageData} from '~/helpers/controller/cms-mixin';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import './bookstore-suppliers.css';

const view = {
    classes: ['bookstore-suppliers', 'page'],
    tag: 'main'
};
const slug = 'pages/print-order';

function providerToModel(p) {
    return {
        name: p.name,
        description: p.blurb || '',
        logoUrl: p.icon,
        buttonUrl: p.url,
        buttonText: p.cta,
        isCanadian: p.canadian
    };
}

function CardFeatures({supplier}) {
    return (
        <React.Fragment>
            <div className="logo-dot">
                <img src={supplier.logoUrl} alt="company logo" />
            </div>
            <h2>{supplier.name}</h2>
            <div className="blurb" dangerouslySetInnerHTML={{__html: supplier.description}} />
            <a className="btn primary" href={supplier.buttonUrl}>{supplier.buttonText}</a>
        </React.Fragment>
    );
}

export function Page() {
    const [pageData, statusPage] = usePageData({slug});

    if (statusPage) {
        return statusPage;
    }

    const suppliers = pageData.providers.map(providerToModel);
    const featuredSupplier = pageData.featured_providers.map(providerToModel); // should only be 1
    const model = {
        headline: pageData.title,
        subhead: pageData.intro_heading,
        subhead2: pageData.intro_description,
        featuredSupplier,
        featuredSuppliersBlurb: pageData.featured_provider_intro_blurb,
        suppliersBlurb: pageData.other_providers_intro_blurb,
        suppliers
    };

    return (
        <React.Fragment>
            <div className="hero">
                <div className="content">
                    <div className="text">
                        <h1>{model.headline}</h1>
                        <div className="small-screen" dangerouslySetInnerHTML={{__html: model.subhead}} />
                        <div className="larger-screen" dangerouslySetInnerHTML={{__html: model.subhead2}} />
                    </div>
                </div>
                <div className="images">
                    <img
                        className="desktop-bg"
                        src="/images/bookstore-suppliers/hero-bookstore-print-desktop.png"
                    />
                </div>
            </div>
            <div className="main-content">
                <div className="text-section">
                    <h2>Providers</h2>
                    <div className="intro-blurb">{model.featuredSuppliersBlurb}</div>
                </div>
                {
                    featuredSupplier.map((supplier) =>
                        <div className="card featured">
                            <CardFeatures supplier={supplier} />
                        </div>
                    )
                }
                <div className="intro-blurb">{model.suppliersBlurb}</div>
                <div className="swipable-row">
                    <div className="cards">
                        {
                            model.suppliers.map((supplier) =>
                                <div className="card">
                                    {
                                        Boolean(supplier.isCanadian) &&
                                            <div className="flag-card">
                                                <div className="canada-flag">
                                                    <div className="white-field">
                                                        <FontAwesomeIcon icon={['fab', 'canadian-maple-leaf']} />
                                                    </div>
                                                </div>
                                            </div>
                                    }
                                    <CardFeatures supplier={supplier} />
                                </div>
                            )
                        }
                    </div>
                </div>
            </div>
        </React.Fragment>
    );
}

export default pageWrapper(Page, view);
