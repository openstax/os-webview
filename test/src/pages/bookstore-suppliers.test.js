import {Page} from '~/pages/bookstore-suppliers/bookstore-suppliers';
import {makeMountRender, snapshotify} from '../../helpers/jsx-test-utils.jsx';

describe('BookstoreSuppliers', () => {
    it('matches snapshot', () => {
        const wrapper = makeMountRender(Page, {})()

        return new Promise((resolve) => {
            setTimeout(() => {
                resolve(expect(snapshotify(wrapper)).toMatchSnapshot());
            }, 0);
        });
    });
});
