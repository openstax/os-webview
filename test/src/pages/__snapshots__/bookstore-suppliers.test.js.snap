// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookstoreSuppliers matches snapshot 1`] = `"<div class=\\"hero\\"><div class=\\"content\\"><div class=\\"text\\"><h1>Need to order print copies for your campus?</h1><div class=\\"small-screen\\">We've got you covered. Check out your print options below.</div><div class=\\"larger-screen\\">If you're an instructor or student and would like to order a print copy of your book, please go to our <b><a href=\\"https://openstax.org/subjects\\">Subjects page</a></b> and select your book to view your options.</div></div></div><div class=\\"images\\"><img class=\\"desktop-bg\\" src=\\"/images/bookstore-suppliers/hero-bookstore-print-desktop.png\\"></div></div>"`;
